### PR TITLE
Add CheckNameAvailability operation to 2017-05-01 Swagger (management plane) for Azure Batch

### PR DESCRIPTION
**Pull request by kiwidev**
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR. 

### Note for reviewers
 - This method follows the RP guidelines from https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/proxy-api-reference.md#check-name-availability-requests

### CODE REVIEW

**Comment by matthchr** [comment]
This looks good to me @kiwidev

**Comment by vishrutshah** [comment]
Thanks @kiwidev for the PR. I'll be reviewing the PR by tomorrow. Feel free to ping me anytime if you need something meanwhile :)

**Comment by ravbhatnagar** [comment]
#ARMSignedOff

File: BatchManagement.json [hunk 1]
```diff
@@ -0,0 +1,1618 @@
...
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Batch/batchAccounts/{accountName}/applications": {
+      "get": {
+        "tags": [
+          "Application"
+        ],
+        "operationId": "Application_List",
```
Comment (vishrutshah) [inline BatchManagement.json]: Should we name it as `Application_ListByBatchAccount` being `applications` as tracked resource?

File: BatchManagement.json [hunk 2]
```diff
@@ -0,0 +1,1618 @@
...
+            "minLength": 3,
+            "maxLength": 24,
+            "description": "A name for the Batch account which must be unique within the region. Batch account names must be between 3 and 24 characters in length and must use only numbers and lowercase letters. This name is used as part of the DNS name that is used to access the Batch service in the region in which the account is created. For example: http://accountname.region.batch.azure.com/."
+          },
+          {
+            "name": "parameters",
+            "in": "body",
+            "required": true,
+            "schema": {
+              "$ref": "#/definitions/BatchAccountCreateParameters"
```
Comment (vishrutshah) [inline BatchManagement.json]: A PUT operation request body schema should be the same as its 200 response schema, to allow reusing the same entity between GET and PUT. If the schema of the PUT request body is a superset of the GET response body, make sure you have a PATCH operation to make the resource updatable. Operation: 'BatchAccount_Create' Request Model: 'BatchAccountCreateParameters' Response Model: 'BatchAccount'

Can you please look why is this not the case?

File: BatchManagement.json [hunk 3]
```diff
@@ -0,0 +1,1618 @@
...
+          "default": {
+            "description": "Error response describing why the operation failed.",
+            "schema": {
+              "$ref": "#/definitions/CloudError"
+            }
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Batch/batchAccounts/{accountName}/applications/{applicationId}/versions/{version}": {
```
Comment (vishrutshah) [inline BatchManagement.json]: Do we not have `ListByApplication` operation for PATH

`/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Batch/batchAccounts/{accountName}/applications/{applicationId}/versions`

Reference: https://travis-ci.org/Azure/azure-rest-api-specs/jobs/254658613#L673

File: BatchManagement.json [hunk 4]
```diff
@@ -0,0 +1,1618 @@
...
+      }
+    }
+  },
+  "paths": {
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Batch/batchAccounts/{accountName}": {
+      "put": {
+        "tags": [
+          "BatchAccount"
+        ],
+        "operationId": "BatchAccount_Create",
```
Comment (vishrutshah) [inline BatchManagement.json]: As we use the `NOUN` part of `NOUN_VERB` pattern in `operationId` in the grouping of operation. Currently, we have a model named as `NOUN` that'll be auto resolved by autorest to resolve naming conflicts. 

Reference: https://travis-ci.org/Azure/azure-rest-api-specs/jobs/254658613#L685

Please make sure that is what you expect in SDK or we need to make them differ here.

File: BatchManagement.json [hunk 5]
```diff
@@ -0,0 +1,1618 @@
...
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Batch/batchAccounts/{accountName}/applications/{applicationId}/versions/{version}/activate": {
+      "post": {
+        "tags": [
+          "ApplicationPackage"
+        ],
+        "operationId": "ApplicationPackage_Activate",
```
Comment (vishrutshah) [inline BatchManagement.json]: As we use the `NOUN` part of `NOUN_VERB` pattern in `operationId` in the grouping of operation. Currently, we have a model named as `NOUN` that'll be auto resolved by autorest to resolve naming conflicts. 

Reference: https://travis-ci.org/Azure/azure-rest-api-specs/jobs/254658613#L916

Please review 18 of these [`OperationIdNounConflictingModelNames`] warning in the build [here](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/254658613#L646)

Please make sure that is what you expect in SDK or we need to make them differ here.

File: BatchManagement.json [hunk 6]
```diff
@@ -0,0 +1,1618 @@
...
+          "description": "A value indicating whether packages within the application may be overwritten using the same version string."
+        },
+        "displayName": {
+          "type": "string",
+          "description": "The display name for the application."
+        }
+      },
+      "description": "Parameters for adding an Application."
+    },
+    "Application": {
```
Comment (vishrutshah) [inline BatchManagement.json]: line 1284: @ravbhatnagar Can you please validate that this is expected?

Top level properties should be one of name, type, id, location, properties, tags, plan, sku, etag, managedBy, identity. Model definition 'Application' has extra properties ['displayName,packages,allowUpdates,defaultVersion']Top level properties should be one of name, type, id, location, properties, tags, plan, sku, etag, managedBy, identity. Model definition 'Application' has extra properties ['displayName,packages,allowUpdates,defaultVersion']

Reference: https://travis-ci.org/Azure/azure-rest-api-specs/jobs/254658613#L1010

File: BatchManagement.json [hunk 7]
```diff
@@ -0,0 +1,1618 @@
...
+          "description": "A value indicating whether packages within the application may be overwritten using the same version string."
+        },
+        "defaultVersion": {
+          "type": "string",
+          "description": "The package to use if a client requests the application but does not specify a version."
+        }
+      },
+      "description": "Contains information about an application in a Batch account."
+    },
+    "ApplicationPackage": {
```
Comment (vishrutshah) [inline BatchManagement.json]: line 1312: @ravbhatnagar same as above. can you please validate?

Top level properties should be one of name, type, id, location, properties, tags, plan, sku, etag, managedBy, identity. Model definition 'ApplicationPackage' has extra properties ['version,state,format,storageUrl,storageUrlExpiry,lastActivationTime']

Reference: https://travis-ci.org/Azure/azure-rest-api-specs/jobs/254658613#L1021

File: BatchManagement.json (hunk 6 above)
Comment (vishrutshah) [inline BatchManagement.json]: line 1284: looks like this is an ARM resource. This should have `id`, `name` & `type` as the `readOnly`. 

Reference: https://travis-ci.org/Azure/azure-rest-api-specs/jobs/254658613#L1045

File: BatchManagement.json [hunk 8]
```diff
@@ -0,0 +1,1618 @@
...
+      "properties": {
+        "id": {
```
Comment (vishrutshah) [inline BatchManagement.json]: If this is a resource then are we missing `name` & `type` as `readOnly` ?

File: BatchManagement.json [hunk 9]
```diff
@@ -0,0 +1,1618 @@
+{
+  "swagger": "2.0",
```
Comment (vishrutshah) [inline BatchManagement.json]: This file is placed at the wrong location. Please move it inside the Batch specifications.

**Review by vishrutshah — CHANGES_REQUESTED** [review]
@kiwidev Left some questions & comments. Let me know what you think. Thanks!

File: BatchManagement.json (hunk 9 above)
Comment (kiwidev) [inline BatchManagement.json]: Oops, this file should not have been included, and has been removed. The diff was only on the BatchManagement.json file inside 2017-05-01. (Only change should be the checkNameAvailability changes).

File: BatchManagement.json (hunk 7 above)
Comment (kiwidev) [inline BatchManagement.json]: This type doesn't match the ARM proxy specification and will be reworked in a future API version.

File: BatchManagement.json (hunk 8 above)
Comment (kiwidev) [inline BatchManagement.json]: This type doesn't match the ARM proxy specification and will be reworked in a future API version.

**Comment by kiwidev** [comment]
@vishrutshah apologies, there was an extra file included in this commit which has now been removed. All of the issues you've pointed out are not changing as part of the PR, and they are mostly all known issues.
As part of the next API version we are planning to revisit the API surface, which will address all / most of these issues. Unfortunately, as these APIs have been in production for a while (in the case of applications since before the ARM spec was well defined for proxy resources) we're stuck with them in these API versions.

The intention of this PR was just to review the "checkNameAvailability" changes.

File: BatchManagement.json (hunk 6 above)
Comment (kiwidev) [inline BatchManagement.json]: This type doesn't match the ARM proxy specification and will be reworked in a future API version.

**Comment by vishrutshah** [comment]
hahaah... thanks for removing that file. Let me review the smaller chunk now :)

**Comment by vishrutshah** [comment]
Closing and re-opening this PR to have the latest CI changes to include CI bug fix.

File: specification/batch/resource-manager/Microsoft.Batch/2017-05-01/BatchManagement.json [hunk 10]
```diff
@@ -1496,6 +1544,44 @@
...
+    },
+    "CheckNameAvailabilityParameters": {
+      "properties": {
+        "name": {
+          "type": "string",
+          "description": "The name to check for availability"
+        },
+        "type": {
+          "type": "string",
+          "description": "The resource type. Must be set to Microsoft.Batch/batchAccounts"
```
Comment (vishrutshah) [inline specification/batch/resource-manager/Microsoft.Batch/2017-05-01/BatchManagement.json]: Should this be modeled as enum with one default value? like: https://github.com/Azure/azure-rest-api-specs/blob/current/specification/storage/resource-manager/Microsoft.Storage/2017-06-01/storage.json#L689 ?

File: specification/batch/resource-manager/Microsoft.Batch/2017-05-01/BatchManagement.json [hunk 11]
```diff
@@ -1496,6 +1544,44 @@
...
+        }
+      },
+      "description": "Parameters for a check name availability request."
+    },
+    "CheckNameAvailabilityResult": {
+      "properties": {
+        "nameAvailable": {
+          "type": "boolean",
```
Comment (vishrutshah) [inline specification/batch/resource-manager/Microsoft.Batch/2017-05-01/BatchManagement.json]: Should this be `readOnly: true` as seems like only settable by server

File: specification/batch/resource-manager/Microsoft.Batch/2017-05-01/BatchManagement.json [hunk 12]
```diff
@@ -1496,6 +1544,44 @@
...
+            "Invalid",
+            "AlreadyExists"
+          ],
+          "x-ms-enum": {
+            "name": "NameAvailabilityReason",
+            "modelAsString": false
+          },
+          "description": "Gets the reason that a Batch account name could not be used. The Reason element is only returned if NameAvailable is false."
+        },
+        "message": {
```
Comment (vishrutshah) [inline specification/batch/resource-manager/Microsoft.Batch/2017-05-01/BatchManagement.json]: `readOnly: true` here as well ?

**Review by vishrutshah — CHANGES_REQUESTED** [review]
@kiwidev Left couple of minor comments. Let me know what you think? Thanks!

File: specification/batch/resource-manager/Microsoft.Batch/2017-05-01/BatchManagement.json (hunk 10 above)
Comment (kiwidev) [inline specification/batch/resource-manager/Microsoft.Batch/2017-05-01/BatchManagement.json]: Changed, thanks.

File: specification/batch/resource-manager/Microsoft.Batch/2017-05-01/BatchManagement.json (hunk 11 above)
Comment (kiwidev) [inline specification/batch/resource-manager/Microsoft.Batch/2017-05-01/BatchManagement.json]: Changed.

File: specification/batch/resource-manager/Microsoft.Batch/2017-05-01/BatchManagement.json (hunk 12 above)
Comment (kiwidev) [inline specification/batch/resource-manager/Microsoft.Batch/2017-05-01/BatchManagement.json]: Changed

**Comment by kiwidev** [comment]
Thanks @vishrutshah, changes made as noted.

**Comment by kiwidev** [comment]
@vishrutshah is the build error on the ruby build: `ValueError: Autorest call ended with 0, but no files were generated` actually a build error or a bug in the CI system that we can work around? Thanks.

**Comment by vishrutshah** [comment]
@kiwidev It's because of the wrong `swagger_to_sdk` config file in Ruby repo. We should be good to move forward here. 

Issue https://github.com/Azure/azure-sdk-for-ruby/issues/853

**Comment by vishrutshah** [comment]
looks like model validation failing but seems like that already existed. 

BatchAccount_Delete: `RESPONSE_STATUS_CODE_NOT_IN_EXAMPLE` Reference: https://travis-ci.org/Azure/azure-rest-api-specs/jobs/255449011#L662

Would you mind fixing this by adding `204` in example or we can open an issue for the service team to follow up. I am fine both the ways. Let me know. Thanks!

**Comment by kiwidev** [comment]
Are you ok with it being done as a separate PR? I'll open up another PR to keep it separate from this.

**Comment by kiwidev** [comment]
@vishrutshah Change made to add 204 to the example, model validation now passing.

**Review by vishrutshah — APPROVED** [review]

**Comment by vishrutshah** [comment]
Thanks @kiwidev for the quick turn-around!  👍

**Pull request merged**